internal/session.go

```diff
@@ -0,0 +1,421 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pborman/uuid"
+)
+
+func init() {
+	RegisterActivityWithOptions(sessionCreationActivity, RegisterActivityOptions{
+		Name: sessionCreationActivityName,
+	})
+	RegisterActivityWithOptions(sessionCompletionActivity, RegisterActivityOptions{
+		Name: sessionCompletionActivityName,
+	})
+}
+
+type (
+	sessionState string
+
+	// SessionInfo contains information of a created session. For now, the only exported
+	// field is SessionID which has string type. SessionID is a uuid generated when
+	// CreateSession() or RecreateSession() is called and can be used to uniquely identify
+	// a session.
+	SessionInfo struct {
+		SessionID    string
+		tasklist     string // tasklist is resource specific
+		sessionState sessionState
+	}
+
+	sessionTokenBucket struct {
+		*sync.Cond
+		availableToken int
+	}
+
+	sessionEnvironment struct {
+		*sync.Mutex
+		doneChanMap              map[string]chan struct{}
+		resourceID               string
+		resourceSpecificTasklist string
+		sessionTokenBucket       *sessionTokenBucket
+		testEnv                  *testWorkflowEnvironmentImpl
+	}
+)
+
+const (
+	sessionStateOpen   sessionState = "open"
+	sessionStateFailed sessionState = "failed"
+	sessionStateClosed sessionState = "closed"
+
+	sessionInfoContextKey        contextKey = "sessionInfo"
+	sessionEnvironmentContextKey contextKey = "sessionEnvironment"
+
+	sessionCreationActivityName   string = "internalSessionCreationActivity"
+	sessionCompletionActivityName string = "internalSessionCompletionActivity"
+
+	errTooManySessionsMsg string = "too many outstanding sessions"
+)
+
+var (
+	errNoOpenSession            = errors.New("no open session in the context")
+	errNoSessionInfo            = errors.New("no session information found in the context")
+	errFoundExistingOpenSession = errors.New("found exisiting open session in the context")
+	errSessionFailed            = errors.New("session has failed")
+)
+
+// Note: Worker should be configured to process session. To do this, set the following
+// fields in WorkerOptions:
+//     EnableSessionWorker: true
+//     SessionResourceID: The identifier of the resource consumed by sessions.
```

Comment:
Add more information on how resourceID is expected to be used. For example for per process cache vs host file system.

Reply:
Leave this as a future work as now we need to hide resourceID from user.